File: 1570-FinalPricesWithASpecialDiscountInAShop/1570-FinalPricesWithASpecialDiscountInAShop.py
# Last updated: 09/07/2026, 10:06:51
class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        stack = []
        answer = prices[:]
        for i in range(len(prices)):
            while stack and prices[i] <= prices[stack[-1]]:
                k = stack.pop()
                answer[k] = prices[k] - prices[i]
            stack.append(i)
        return answer     


        

File: 1570-FinalPricesWithASpecialDiscountInAShop/1570-FinalPricesWithASpecialDiscountInAShop.java
// Last updated: 09/07/2026, 10:06:47
class Solution {
    public int[] finalPrices(int[] prices) {
        Stack<Integer> s = new Stack<>();
        int n = prices.length;
        for(int i = n-1;i>=0;i--){
            int cur = prices[i];
            while(!s.isEmpty() && s.peek() > cur){
                
                s.pop();

            }
            if(!s.isEmpty()) {
                prices[i] = prices[i] - s.peek();
            }
            s.push(cur);
        }
        return prices;
    }
}